Prevent double 409 conflict form errors after saving the work package

oliverguenther opened the pull request:
Due to the way the changesets are used and cleared, we were causing
double posting to the form with an old work package lockVersion:

1. Lock version in changeset correct, editing work package, saving it

2. The EditForm destroys the changeset on submit prior to the work
package being updated in state

3. The WPStatusButtonComponent got a call after the changeset was
destroyed, resulting in an immediate new changeset being created.

4. The new changeset resulted in the changeset set being fired, which
triggered the select edit fields

5. They requested the form post with an older lock version

Without 3.), we don't always have a changeset present for the resource, so we need to rely on the `temporaryEditResource` to get either the projected or the pristine work package in other places (such as the single-view). Otherwise, switching types will not properly reload the form

oliverguenther committed:
Prevent double 409 conflict form errors after saving the work package

Due to the way the changesets are used and cleared, we were causing
double posting to the form with an old work package lockVersion:

1. Lock version in changeset correct, editing work package, saving it

2. The EditForm destroys the changeset on submit prior to the work
package being updated in state

3. The WPStatusButtonComponent got a call after the changeset was
destroyed, resulting in an immediate new changeset being created.

4. The new changeset resulted in the changeset set being fired, which
triggered the select edit fields

5. They requested the form post with an older lock version

openprojectci commented on frontend/src/app/components/wp-buttons/wp-status-button/wp-status-button.component.ts:
```diff
@@ -35,6 +35,7 @@ import {Highlighting} from "core-components/wp-fast-table/builders/highlighting/
 import {WorkPackageCacheService} from "core-components/work-packages/work-package-cache.service";
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {WorkPackageChangeset} from "core-components/wp-edit/work-package-changeset";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/3e56c00ac126880e54f52ed25db2d574d6e8b35a/frontend/src/app/components/wp-buttons/wp-status-button/wp-status-button.component.ts#L39"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts:
```diff
@@ -43,6 +43,7 @@ import {IFieldSchema} from "core-app/modules/fields/field.base";
 import {EditingPortalService} from "core-app/modules/fields/edit/editing-portal/editing-portal-service";
 import {EditFormRoutingService} from "core-app/modules/fields/edit/edit-form/edit-form-routing.service";
+import {ResourceChangesetCommit} from "core-app/modules/fields/edit/services/hal-resource-editing.service";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/890e8b37ca5a7cceafb46aaf43dc81b7c245c18c/frontend/src/app/modules/fields/edit/edit-form/edit-form.component.ts#L47"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

openprojectci commented on frontend/src/app/components/wp-buttons/wp-status-button/wp-status-button.component.ts:
```diff
@@ -35,6 +35,8 @@ import {Highlighting} from "core-components/wp-fast-table/builders/highlighting/
 import {UntilDestroyedMixin} from "core-app/helpers/angular/until-destroyed.mixin";
+import {concatAll} from "rxjs/operators";
+import {combineLatest} from "rxjs";
 
```
<table data-meta="generated_by_danger">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><del><span data-href="https://github.com/opf/openproject/blob/890e8b37ca5a7cceafb46aaf43dc81b7c245c18c/frontend/src/app/components/wp-buttons/wp-status-button/wp-status-button.component.ts#L40"></span>Please use <code>ChangeDetectionStrategy.OnPush</code> for this component</del></td>
    </tr>
  </tbody>
</table>

oliverguenther committed:
Ensure we're not passing a proxy resource as pristineResource

Since the proxy is indistinguishable from a work package, if you try to
update a work package state with the `projectedResource`, you will end
up with a proxy object in a change, causing stack size errors.

HDinger reviewed (approved):
👍
